v2/blockstore/readwrite.go

```diff
@@ -59,48 +65,187 @@ func WithIndexPadding(p uint64) Option {
+		return err
 	}
-	if err := carv1.WriteHeader(v1Header, b.carV1Writer); err != nil {
-		return nil, fmt.Errorf("couldn't write car header: %w", err)
+	if version != 2 {
+		// The file is not a CAR v2 and we cannot resume from it.
+		return fmt.Errorf("cannot resume on CAR file with version %v", version)
 	}
-	return b, nil
+
+	// Check if file is finalized.
+	// A file is finalized when it contains a valid CAR v2 header with non-zero v1 offset.
+	// If finalized, cannot resume from.
+	var headerInFile carv2.Header
+	if _, err := headerInFile.ReadFrom(internalio.NewOffsetReadSeeker(b.f, carv2.PragmaSize)); err == nil {
+		// TODO: we technically can; this could be a feature to do if asked for.
+		if headerInFile.CarV1Offset != 0 {
+			return fmt.Errorf("cannot resume from a finalized file")
+		}
+	}
+
+	// Use the given CAR v1 padding to instantiate the CAR v1 reader on file.
+	v1r := internalio.NewOffsetReadSeeker(b.ReadOnly.backing, 0)
+	header, err := carv1.ReadHeader(bufio.NewReader(v1r))
+	if err != nil {
+		// Cannot read the CAR v1 header; the file is most likely corrupt.
+		return fmt.Errorf("error reading car header: %w", err)
+	}
+	if !header.Equals(carv1.CarHeader{Roots: roots, Version: 1}) {
+		// Cannot resume if version and root does not match.
+		return errors.New("cannot resume on file with mismatching data header")
+	}
+
+	// TODO See how we can reduce duplicate code here.
+	// The code here comes from index.Generate.
+	// Copied because we need to populate an insertindex, not a sorted index.
+	// Producing a sorted index via generate, then converting it to insertindex is not possible.
+	// Because Index interface does not expose internal records.
+	// This may be done as part of https://github.com/ipld/go-car/issues/95
+
+	offset, err := carv1.HeaderSize(header)
+	if err != nil {
+		return err
+	}
+	frameOffset := int64(0)
+	if frameOffset, err = v1r.Seek(int64(offset), io.SeekStart); err != nil {
+		return err
+	}
+
+	for {
+		// Grab the length of the frame.
+		// Note that ReadUvarint wants a ByteReader.
+		length, err := varint.ReadUvarint(v1r)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+
+		// Null padding; treat zero-length frames as an EOF.
+		// They don't contain a CID nor block, so they're not useful.
+		if length == 0 {
+			break // TODO This behaviour should be an option, not default. By default we should error. Hook this up to a write option
+		}
+
+		// Grab the CID.
+		n, c, err := cid.CidFromReader(v1r)
+		if err != nil {
+			return err
+		}
+		b.idx.InsertNoReplace(c, uint64(frameOffset))
+
+		// Seek to the next frame by skipping the block.
+		// The frame length includes the CID, so subtract it.
+		if frameOffset, err = v1r.Seek(int64(length)-int64(n), io.SeekCurrent); err != nil {
+			return err
+		}
+	}
+	// Seek to the end of last skipped block where the writer should resume writing.
+	_, err = b.carV1Writer.Seek(frameOffset, io.SeekStart)
```

Comment:
@masih resumption should be used with deduplicated puts, so puts that got resumed don't get written again.

We can add this to the doc and or warn people about it unless they explicitly want duplicate blocks in there.